src/models/document/document-content.ts

```diff
@@ -23,7 +30,89 @@ import { DocumentContentModelWithAnnotations } from "./document-content-with-ann
  * Note: the name "DocumentContent" is important because it is used in other
  * parts of the code to find a MST parent with this name.
  */
-export const DocumentContentModel = DocumentContentModelWithAnnotations.named("DocumentContent");
+export const DocumentContentModel = DocumentContentModelWithAnnotations.named("DocumentContent")
+.views(self => ({
+  snapshotWithUniqueIds(asTemplate = false) {
+    const snapshot = cloneDeep(getSnapshot(self));
+    const tileIdMap: { [id: string]: string } = {};
+    const annotationIdMap: { [id: string]: string } = {};
+
+    snapshot.tileMap = (tileMap => {
+      const _tileMap: { [id: string]: ITileModelSnapshotOut } = {};
+      each(tileMap, (tile, id) => {
+        tileIdMap[id] = tile.id = uniqueId();
+        _tileMap[tile.id] = tile;
+      });
+      return _tileMap;
+    })(snapshot.tileMap);
+
+    // Update the sharedModels with new tile ids
+    each(snapshot.sharedModelMap, (sharedModelEntry, id) => {
+      const _tiles = cloneDeep(sharedModelEntry.tiles);
+      sharedModelEntry.tiles = [];
+      _tiles.forEach(tile => {
+        sharedModelEntry.tiles.push(tileIdMap[tile]);
+      });
+      // Update references to provider
+      if (sharedModelEntry.provider) {
+        sharedModelEntry.provider = tileIdMap[sharedModelEntry.provider];
+      }
+      const sharedModel = sharedModelEntry.sharedModel;
+      // TODO: Figure out why linting is failing
+      // if ("providerId" in sharedModel && typeof sharedModel.providerId === "string") {
+      //   sharedModel.providerId = tileIdMap[sharedModel.providerId];
+      if ("providerId" in sharedModel) {
+        const providedSharedModel = sharedModel as any;
+        if (typeof providedSharedModel.providerId === "string") {
+          providedSharedModel.providerId = tileIdMap[providedSharedModel.providerId];
+        }
+      }
+    });
+    // TODO: Give the shared models new ids
+
+    // Update annotations with new tile ids
+    each(snapshot.annotations, (annotation, id) => {
```

Comment:
It is probably another story, but I'd guess that drag and drop needs separate work. I thought they shared the id updating code, but based on this it doesn't seem like it. 
Perhaps it is in that code where we let tiles and shared models update ids within their own content. That will be challenging for the annotation system if those internal id changes also have to be applied to the annotation ids.

Reply:
I've been working on copying sparrows in various ways in another PR that'll come after this one. It's mostly complete with just a couple of lingering issues to deal with. So far it seems like tiles generally don't change their internal ids, except shared data sets update their case and attribute ids.